pkg/tiller/server.go

```diff
@@ -41,29 +53,97 @@ func NewServer() *grpc.Server {
 	)
 }
 
+func authenticate(ctx context.Context) (context.Context, error) {
+	md, ok := metadata.FromContext(ctx)
+	if !ok {
+		return nil, errors.New("Missing metadata in context.")
+	}
```

Comment:
This seems really complicated for what you are trying to do. Why not construct a single client config object and just run a `client.DiscoveryClient.ServerVersion()`? Wouldn't that just let the client library do most of what you do in the multiple functions below? This would be even easier if you sent a sanitized client config instead of storing various bits of the client config in different gRPC headers.